Add python code to visualize Dat files

nilsdeppe opened the pull request:
## Proposed changes

Python script that visualizes Dat files from HDF5 files, and can also dump them as text/ASCII dat/tsv files.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add python code to visualize Dat files

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+    def available_subfiles(h5file):
+        return [member for member in h5file.keys() if member.endswith('.dat')]
+
+    with h5py.File(filename) as h5file:
```
I get this warning in both python2 and python3 with h5py==2.10.0:
```
PlotDatFile.py:116: H5pyDeprecationWarning: The default file mode will change to 'r' (read-only) in h5py 3.0. To suppress this warning, pass the mode you need to h5py.File(), or set the global default h5.get_config().default_file_mode, or set the environment variable H5PY_DEFAULT_READONLY=1. Available modes are: 'r', 'r+', 'w', 'w-'/'x', 'a'. See the docs for details.
```

For python3, if I use h5py==3.7.0, this error goes away. The last version of h5py that works with python2 is 2.10.0, though.

I suggest just explicitly using read mode like

```python
with h5py.File(filename, 'r') as h5file:
```

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+                    "Unable to open dat file '%s'. Available files are:\n%s" %
+                    (subfile_name + '.dat', available_subfiles(h5file)))
+            return process_subfile(dat_file, **kwargs)
+        else:
+            print("Available dat files are:", available_subfiles(h5file))
+            return available_subfiles(h5file)
```
In python3 this prints fine:
```
Available dat files are: ['AhA.dat', 'AhB.dat', 'Norms.dat']
```

However, in python2 it looks like (without the warning from h5py that's raised above in line 116):

```
('Available dat files are:', [u'AhA.dat', u'AhB.dat', u'Norms.dat'])
```

I wonder if there's a nicer way to format things? At the very least maybe use a different encoding when printing to avoid the Unicode "u" before the string?

nilsdeppe replied:
My experience with dealing with string encoding in python has been terrible... Even python 3 made several changes between versions. I'd prefer to just ignore it here since python 2 is EOL and while not pretty it's not a functional deficit. I have no objection to the change as a whole, just not sure it's worth the effort.

knelli2 replied:
I'm ok with leaving as is. Yeah it was mostly a cosmetic change, but I agree all the necessary info is there.

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+            return process_subfile(dat_file, **kwargs)
+        else:
+            print("Available dat files are:", available_subfiles(h5file))
+            return available_subfiles(h5file)
```
This only prints `dat` files in the root group of the H5 file. Any `dat` files inside nested groups like `/MemoryMonitors/Interpolator.dat` or `/ControlSystems/Rotation/x.dat` aren't shown. Maybe change `available_subfiles()` to look recursively through the H5 tree?

knelli2 replied:
Maybe add a little more help text to indicate that you need to provide a `--subfile-name` argument to actually plot things like

```python
print("No subfile was specified. If you want to plot or write data, you need to specify a dat file using the --subfile-name option.")
print("Available dat files are:", available_subfiles(h5file))
```

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+    legend = [(x.decode('ascii') if type(x) == bytes else x)
+              for x in dat_file.attrs['Legend']]
+    if legend_only:
+        print("The legend in the dat subfile is:\n{}".format(legend))
```
Again for python2, the strings printed out have the Unicode "u" in front of them.

knelli2 replied:
Maybe also add in the dat file name when printing out so the user remembers which file they were getting the legend of (because honestly I forgot in 0.1s...) like

```python
print("The legend in the dat subfile '{}' is:\n{}".format(dat_file.name, legend))
```

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+        plt.ylim(float(args['y_bounds'][0]), float(args['y_bounds'][1]))
+    if args['title'] != None:
+        plt.title(args['title'], fontsize=args['fontsize'])
+    plt.savefig(args['output'] + '.pdf',
```
Maybe print something here before you save the figure like you do for `--legend-only` and `--write-dat` like

```python
print("Saving figure to '{}'".format(args['output'] + '.pdf'))
```

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+                            required=False,
+                            nargs=2,
+                            help="The lower and upper bounds of the y-axis.")
+        parser.add_argument('--x-label',
+                            required=False,
+                            help="The label on the x-axis. "
+                            "Default is the name of the x-axis entry.")
```
What do you think about adding a `--y-label` option as well? For just one plot I can see how this could be sort of redundant with the label, but if you include many functions on one plot, it may be nice to label the y-axis with something common to all the functions. Default could just be no y-label.

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+            "the functions to specified labels. An example of what\n"
+            "to pass to the argument (including the single quotes)\n"
+            "is:\n"
+            "'{\"Error(DivergenceCleaningField)\":\"$L_2(\\\\Phi)$\",\n"
+            "\"Error(RestMassDensity)\":\"$L_2(\\\\mathcal{E}(\\\\rho))$\"}'")
```
Also make a note that if you are using LaTeX and you want to do something like `\mathcal{E}`, that you'll actually need to write `\\mathcal{E}`. I didn't do this, just wrote `--labels '{"Size on node 0 (MB)":"$\mathcal{E}$"}'` and got an error:

```
PlotDatFile.py: error: argument --labels: invalid loads value: '{"Size on node 0 (MB)":"$\\mathcal{E}$"}'
```

And then if I did `--labels '{"Size on node 0 (MB)":"$\beta$"}'` instead of `\\beta` I got a really long ugly traceback that ended in:

```
  File "/home/knelli/.local/lib/python2.7/site-packages/matplotlib/backends/backend_pdf.py", line 2153, in get_text_width_height_descent
    self.mathtext_parser.parse(s, 72, prop)
  File "/home/knelli/.local/lib/python2.7/site-packages/matplotlib/mathtext.py", line 3293, in parse
    box = self._parser.parse(s, font_output, fontsize, dpi)
  File "/home/knelli/.local/lib/python2.7/site-packages/matplotlib/mathtext.py", line 2521, in parse
    six.text_type(err)]))
ValueError: 
eta$
^
Expected end of text, found u'$'  (at char 0), (line:1, col:1)
```

Another idea is to just `try; catch` an exception when you're plotting and just print out something like "Unable to format labels. Make sure any LaTeX expressions have two `\\` instead of one"

nilsdeppe replied:
I'll add a note to the help string. Note that the help string already prints out with double slashes

knelli2 replied:
Yeah I saw that and thought it was a typo so I ignored it when testing....learned my lesson! The note you added is helpful.

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+
+    # Make sure the correct number of line styles were supplied
+    if num_linestyles != num_functions and num_linestyles != 1:
+        parser.error(
+            "You must supply either zero, one, or one line style per "
+            "function but parsed {} line styles with {} functions.\n".format(
+                num_linestyles, num_functions))
```
I supplied zero arguments to the `--linestyles` option and got an error. I think this should be reworded to "either one linestyle or one linestyle per function".

Also the second half of the sentence doesn't really make sense with the first. How about

```python
parser.error(
            "You must supply either one line style or one line style per "
            "function. However, {} line styles with {} functions were parsed.\n".format(
                num_linestyles, num_functions))
```

knelli2 reviewed (changes requested):
I've tested all the options for this with python 2.7.18 and python 3.8.10 and everything works fine (minus the comments I made below).

Do you think it's worth putting this into the `bin` in the build directory along with GenerateXdmf and Render1D?

nilsdeppe commented:
I've pushed a fixup.

On my machine at least this is already being placed in `bin/` as `PlotDatFile`

knelli2 commented on tests/Unit/Visualization/Python/Test_PlotDatFile.py:
```diff
@@ -90,6 +91,157 @@ def test_write_plot(self):
+                                       y_label="Blah",
+                                       title="DatTestdata",
+                                       output=plot_filename)
+        # We can't test that the plot is correct, so just make sure it at least
+        # gets written out.
+        self.assertTrue(os.path.exists(plot_filename + '.pdf'))
```
This test seems to be timing out on multiple CI builds. Possibly because you're saving (and removing) 6 additional figures?

knelli2 commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,296 @@
+        '-o',
+        help="Name of the plot file. The '.pdf' extension will "
+        "be added automatically. Note: Requires additional arguments to be "
+        "specified. Specify --help alongside '--output' or '-o' to see "
+        "additional arguments.")
```
I didn't catch this before, but when I do
```
python PlotDatFile.py --help -o
```

I get the expected help message. However, if I do

```
python PlotDatFile.py -o --help
```

I get this error message:

```
PlotDatFile.py: error: argument --output/-o: expected one argument
```

Is there a way to parse this so that it looks for `--help` first before it parses `-o` so we avoid this error?

nilsdeppe replied:
I have no control over the parser. This just uses python's standard parser

knelli2 reviewed (changes requested):
Looks good. You can squash in the existing changes.

Missed something and noticed the test is timing out often on CI.

nilsdeppe commented:
@knelli2 I've squashed, rebased, and increased the test timeout. I increased the timeout since I'd like to avoid these scripts breaking in the future and I have no idea how to test the arguments and calls to matplotlib without writing a file to disk :(

kidder commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,318 @@
+        '-o',
+        help="Name of the plot file. The '.pdf' extension will "
+        "be added automatically. Note: Requires additional arguments to be "
+        "specified. Specify --help alongside '--output' or '-o' to see "
```
alongside -> followed by

kidder commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,318 @@
+        help="Name of the plot file. The '.pdf' extension will "
+        "be added automatically. Note: Requires additional arguments to be "
+        "specified. Specify --help alongside '--output' or '-o' to see "
+        "additional arguments.")
```
"the additional arguments"

kidder commented on src/Visualization/Python/PlotDatFile.py:
```diff
@@ -0,0 +1,318 @@
+        parser.add_argument('--x-axis',
+                            required=False,
+                            help="If specified, make the range be the "
+                            "values is this field of the subfile. Defaults to "
```
values in this

kidder reviewed (changes requested):
Also document the existence of this somewhere (maybe `docs/Tutorials/Visualization.md`)

nilsdeppe commented:
Thanks for the reviews! I've squashed in the changes and rebased on develop. Cheers!

nilsdeppe commented:
CI failure is unrelated to this PR
